File: src/jvm/main/org/jetbrains/kotlinx/lincheck/paramgen/IntGen.java
/*
 * Lincheck
 *
 * Copyright (C) 2019 - 2023 JetBrains s.r.o.
 *
 * This Source Code Form is subject to the terms of the
 * Mozilla Public License, v. 2.0. If a copy of the MPL was not distributed
 * with this file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

package org.jetbrains.kotlinx.lincheck.paramgen;

import org.jetbrains.kotlinx.lincheck.RandomProvider;

public class IntGen implements ParameterGenerator<Integer> {

    private final ExpandingRangeIntGenerator generator;

    public IntGen(RandomProvider randomProvider, String configuration) {
        generator = ExpandingRangeIntGenerator.create(randomProvider.createRandom(), configuration, Integer.MIN_VALUE, Integer.MAX_VALUE, "int");
    }

    public Integer generate() {
        return generator.nextInt();
    }

    @Override
    public void resetRange() {
        generator.restart();
    }
}

File: src/common/main/org/jetbrains/kotlinx/lincheck/paramgen/IntGen.kt
/*
 * Lincheck
 *
 * Copyright (C) 2019 - 2020 JetBrains s.r.o.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Lesser Public License for more details.
 *
 * You should have received a copy of the GNU General Lesser Public
 * License along with this program.  If not, see
 * <http://www.gnu.org/licenses/lgpl-3.0.html>
 */
package org.jetbrains.kotlinx.lincheck.paramgen

import kotlin.random.Random

class IntGen(configuration: String) : ParameterGenerator<Int> {
    private val random = Random(0)
    private var begin = 0
    private var end = 0

    init {
        if (configuration.isEmpty()) { // use default configuration
            begin = DEFAULT_BEGIN
            end = DEFAULT_END
        } else {
            val args = configuration.replace("\\s".toRegex(), "").split(":".toRegex()).toTypedArray()
            when (args.size) {
                2 -> {
                    begin = args[0].toInt()
                    end = args[1].toInt()
                }
                else -> throw IllegalArgumentException("Configuration should have " +
                        "two arguments (begin and end) separated by colon")
            }
        }
    }

    override fun generate(): Int {
        return begin + random.nextInt(end - begin + 1)
    }

    fun checkRange(min: Int, max: Int, type: String) {
        require(!(begin < min || end - 1 > max)) {
            ("Illegal range for "
                    + type + " type: [" + begin + "; " + end + ")")
        }
    }
}

private const val DEFAULT_BEGIN = -10
private const val DEFAULT_END = 10